driver/src/orderbook/streamed/state.rs

```diff
@@ -0,0 +1,561 @@
+                    buy_token: order.buy_token,
+                    sell_token: order.sell_token,
+                    buy_amount,
+                    sell_amount,
+                }
+            })
+    }
+
+    /// Reset the state to the default state in which no events have been applied.
+    pub fn clear(&mut self) {
+        self.orders.clear();
+        self.balances.clear();
+        self.tokens.0.clear();
+    }
+
+    /// Apply an event to the state, modifying it.
+    ///
+    /// In case of error no modifications take place.
+    ///
+    /// `block_batch_id` is the current batch based on the timestamp of the block that contains the
+    ///  event.
+    pub fn event(&mut self, event: &Event, block_batch_id: BatchId) -> Result<(), Error> {
+        match event {
+            Event::Deposit(event) => self.deposit(event, block_batch_id),
+            Event::WithdrawRequest(event) => self.withdraw_request(event),
+            Event::Withdraw(event) => self.withdraw(event, block_batch_id),
+            Event::TokenListing(event) => self.token_listing(event),
+            Event::OrderPlacement(event) => self.order_placement(event),
+            Event::OrderCancellation(event) => self.order_cancellation(event, block_batch_id),
+            Event::OrderDeletion(event) => self.order_deletion(event),
+            Event::Trade(event) => unimplemented!(),
+            Event::TradeReversion(event) => unimplemented!(),
+            Event::SolutionSubmission(event) => self.solution_submission(event, block_batch_id),
+        }
+    }
+
+    fn deposit(&mut self, event: &Deposit, block_batch_id: BatchId) -> Result<(), Error> {
+        self.balances
+            .entry((event.user, event.token))
+            .or_default()
+            .deposit(
+                Flux {
+                    amount: event.amount,
+                    batch_id: event.batch_id,
+                },
+                block_batch_id,
+            );
+        Ok(())
+    }
+
+    fn withdraw_request(&mut self, event: &WithdrawRequest) -> Result<(), Error> {
+        let balance = self.balances.entry((event.user, event.token)).or_default();
+        balance.pending_withdraw = Some(Flux {
+            amount: event.amount,
+            batch_id: event.batch_id,
+        });
+        Ok(())
+    }
+
+    fn withdraw(&mut self, event: &Withdraw, block_batch_id: BatchId) -> Result<(), Error> {
+        let balance = self.balances.entry((event.user, event.token)).or_default();
+        balance.update_deposit_balance(block_batch_id);
+        balance.pending_withdraw = None;
+        balance.balance = balance
+            .balance
+            .checked_sub(event.amount)
+            .ok_or(Error::MathUnderflow)?;
+        Ok(())
+    }
+
+    fn token_listing(&mut self, event: &TokenListing) -> Result<(), Error> {
+        self.tokens.0.push((event.id, event.token));
+        Ok(())
+    }
+
+    fn order_placement(&mut self, event: &OrderPlacement) -> Result<(), Error> {
```

Comment:
naming nit: maybe these can all be imperatives? E.g. place_order/request_withdraw or process_order_placement if you prefer keeping the event name.

Reply:
I don't mind but is it really an improvement? These functions are named exactly like the event type they handle but in snake_case.

Reply:
For consistency with `apply_event`, it could be `apply_order_placement`.

Reply:
I don't feel strongly either way.

Reply:
I'll rename it to `apply_` in the next PR (will make merging easier).